### PR TITLE
Implement health and fall damages

### PR DESCRIPTION
**Pull request by Sceat**

### CODE REVIEW

**Comment by Sceat** [comment]
seems that i missed something with the `dispatch` :thinking: i never get my action

```js
log.debug({ fall_distance, damage }, 'has fallen')
if (damage > 0) dispatch('fall_damage', { damage })
```

```
packet/position
packet/position
packet/position
packet/position
packet/position
packet/position
packet/position
packet/position
packet/position
0000033543 player/fall_dama  D  has fallen { fall_distance: 4.919999986886978, damage: 0.9599999934434891 }
packet/keep_alive
packet/position
packet/position
packet/position
packet/position
packet/position
packet/position
packet/keep_alive
packet/position
```

**Comment by Sceat** [comment]
seems to be caused by the combine async iterator lib 🤡

**Comment by DeltaEvo** [comment]
Oups, my bad, combineAsyncIteratot expect to have async iterator and actions is an async iterable so it takes it as the options object, to fix that we can do
```js
combineAsyncIterators(actions[Symbol.asyncIterator](), packets)
```

**Comment by Sceat** [comment]
PR is kinda done, just need to figure out why
```js
client.write('update_health', {
          health,
          food: 20,
          foodSaturation: 0,
        })
```
doesn't update the client health bar

**Comment by Sceat** [comment]
for whatever reason we cannot send more than `20` of health :thinking:

**Comment by DeltaEvo** [comment]
Maybe you should send a https://wiki.vg/Protocol#Entity_Properties to updqte it's max health before

**Comment by Sceat** [comment]
okay well i forgot everything about this game damn me

**Comment by Sceat** [comment]
image

File: src/player/fall_damage.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import { aiter } from 'iterator-helper'
+
+import logger from '../logger.js'
+
+const log = logger(import.meta)
+
+export default {
+  reducer(state, { type, payload }) {
+    if (type !== 'fall_damage') return state
```
Comment (DeltaEvo) [inline src/player/fall_damage.js:11]: Can you invert the condition and the branches ?

File: src/position.js [hunk 2]
```diff
@@ -1,17 +1,14 @@
...
-  if (types.includes(type)) {
-    const { onGround, ...position } = payload
-    return {
-      ...state,
-      position: {
-        ...state.position,
-        ...position,
-      },
-    }
+  if (!types.includes(type)) return state
```
Comment (DeltaEvo) [inline src/position.js:4]: Can you reinvert the condition ?

File: src/player/health.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    const reducer = (last_health, [{ health }]) => {
+      if (last_health !== health) {
+        client.write('update_health', {
+          health,
+          food: 20,
+          foodSaturation: 0.0,
+        })
+      }
+      return health
+    }
```
Comment (DeltaEvo) [inline src/player/health.js:16]: Can you inline this reducer ?

**Review by DeltaEvo — CHANGES_REQUESTED** [review]
Just some little things but seems good otherwise

**Review by DeltaEvo — APPROVED** [review]

**Pull request merged**